bitcoin/src/bip32.rs

```diff
@@ -256,108 +368,272 @@ impl serde::Serialize for ChildNumber {
         u32::from(*self).serialize(serializer)
     }
 }
+/// `normal_derivation_path` module with inner methods.
+pub mod normal_derivation_path {
+    use super::*;
 
-/// Trait that allows possibly failable conversion from a type into a
-/// derivation path
-pub trait IntoDerivationPath {
-    /// Converts a given type into a [`DerivationPath`] with possible error
-    fn into_derivation_path(self) -> Result<DerivationPath, Error>;
-}
+    /// A [`DerivationPath`] that can only contain [`NormalChildNumber`] (i.e. no hardened indexes).
+    #[derive(Clone, PartialEq, Eq, Debug, PartialOrd, Ord, Hash)]
+    pub struct NormalDerivationPath(Vec<NormalChildNumber>);
 
-/// A BIP-32 derivation path.
-#[derive(Clone, PartialEq, Eq, Ord, PartialOrd, Hash)]
-pub struct DerivationPath(Vec<ChildNumber>);
+    #[cfg(feature = "serde")]
+    internals::serde_string_impl!(
+        NormalDerivationPath,
+        "a BIP-32 derivation path containing normal numbers only"
+    );
+
+    impl NormalDerivationPath {
+        /// The Master (empty) path (i.e. the master key).
+        #[allow(dead_code)]
+        const MASTER: Self = NormalDerivationPath(Vec::new());
+
+        /// Creates a new [`NormalDerivationPath`] from a vector of [`NormalChildNumber`]s.
+        pub fn new(vec: Vec<NormalChildNumber>) -> Self { NormalDerivationPath(vec) }
+        /// Returns length of the [`NormalDerivationPath`].
+        pub fn len(&self) -> usize { self.0.len() }
+        /// Returns a [`DerivationPath`] from the [`NormalDerivationPath`].
+        pub fn to_derivation_path(&self) -> DerivationPath {
+            self.into_iter().map(|n| n.to_number()).collect()
+        }
 
-#[cfg(feature = "serde")]
-internals::serde_string_impl!(DerivationPath, "a BIP-32 derivation path");
+        /// Returns `true` if the [`NormalDerivationPath`] is empty.
+        pub fn is_empty(&self) -> bool { self.0.is_empty() }
 
-impl<I> Index<I> for DerivationPath
-where
-    Vec<ChildNumber>: Index<I>,
-{
-    type Output = <Vec<ChildNumber> as Index<I>>::Output;
+        /// Returns whether [`NormalDerivationPath`] represents master key (i.e. it's length
+        /// is empty). True for `m` path.
+        pub fn is_master(&self) -> bool { self.is_empty() }
 
-    #[inline]
-    fn index(&self, index: I) -> &Self::Output { &self.0[index] }
-}
+        /// Creates a new [`NormalDerivationPath`] that is a child of this one.
+        pub fn child(&self, cn: NormalChildNumber) -> Self { self.new_extended_from([cn]) }
 
-impl Default for DerivationPath {
-    fn default() -> DerivationPath { DerivationPath::master() }
-}
+        /// Converts into a [`NormalDerivationPath`] that is a child of this one.
+        pub fn into_with_child(self, cn: NormalChildNumber) -> Self {
```

Comment:
I don't love the name but I can't think of any better. But why not just have `fn push_child(&mut self, cn: ChildNumber)`?

Reply:
i did not understand the point of this comment, do you want a rename ?

Reply:
I suggest changing both the interface and the name.